pkg/process/checks/chunking_test.go

```diff
@@ -0,0 +1,622 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package checks
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	model "github.com/DataDog/agent-payload/v5/process"
+	"github.com/stretchr/testify/assert"
+)
+
+type testPayload struct {
+	id     int
+	weight int
+}
+
+type testPayloadList struct {
+	payloads []*testPayload
+	chunker  testPayloadChunker
+}
+
+type testPayloadChunker interface {
+	Accept(payloads []*testPayload, weight int)
+}
+
+func (l *testPayloadList) Len() int {
+	return len(l.payloads)
+}
+
+func (l *testPayloadList) WeightAt(idx int) int {
+	if idx >= len(l.payloads) {
+		return 0
+	}
+	return l.payloads[idx].weight
+}
+
+func (l *testPayloadList) ToChunk(start, end int, weight int) {
+	l.chunker.Accept(l.payloads[start:end], weight)
+}
+
+type testChunker struct {
+	chunkPropsTracker
+	chunks [][]*testPayload
+}
+
+func (c *testChunker) Accept(payloads []*testPayload, weight int) {
+	if c.idx >= len(c.chunks) {
+		c.chunks = append(c.chunks, []*testPayload{})
+		c.props = append(c.props, chunkProps{})
+	}
+
+	c.chunks[c.idx] = append(c.chunks[c.idx], payloads...)
+	c.props[c.idx].size += len(payloads)
+	c.props[c.idx].weight += weight
+}
+
+type chunkGroup struct {
+	weights []int
+	start   int
+}
+
+type chunkTest struct {
+	maxChunkWeight int
+	maxChunkSize   int
+	groups         []chunkGroup
+
+	expectedIDs [][]int
+}
+
+func (ct *chunkTest) runGroup(id int, chunker *testChunker, g chunkGroup) {
+	payloads := make([]*testPayload, len(g.weights))
+	for i := range payloads {
+		payloads[i] = &testPayload{
+			id:     id,
+			weight: g.weights[i],
+		}
+		id++
+	}
+
+	list := &testPayloadList{
+		payloads: payloads,
+		chunker:  chunker,
+	}
+
+	chunker.idx = g.start
+	chunkPayloadsBySizeAndWeight(list, chunker, ct.maxChunkSize, ct.maxChunkWeight)
+}
+
+func (ct *chunkTest) run(t *testing.T) {
+	t.Helper()
+	chunker := &testChunker{}
+
+	id := 1
+	for _, g := range ct.groups {
+		ct.runGroup(id, chunker, g)
+		id += len(g.weights)
+	}
+	actualIDs := make([][]int, len(chunker.chunks))
+	for i := range chunker.chunks {
+		for _, p := range chunker.chunks[i] {
+			actualIDs[i] = append(actualIDs[i], p.id)
+		}
+	}
+
+	assert.Equal(t, ct.expectedIDs, actualIDs)
+}
+
+func TestChunkPayloadsBySizeAndWeightSingleRun(t *testing.T) {
+	tests := []chunkTest{
+		{
+			maxChunkWeight: 4,
+			maxChunkSize:   3,
+			groups: []chunkGroup{
+				{
+					weights: []int{
+						5,
+						// chop!
+						1,
+						1,
+						1,
+						// chop!
+						1,
+					},
+				},
+			},
+			expectedIDs: [][]int{
+				{
+					1,
+				},
+				{
+					2, 3, 4,
+				},
+				{
+					5,
+				},
+			},
+		},
+		{
+			maxChunkWeight: 4,
+			maxChunkSize:   3,
+			groups: []chunkGroup{
+				{
+					weights: []int{
+						1,
+						3,
+						// chop!
+						1,
+						1,
+						1,
```

Comment:
is it a `// chop!` comment missing here?

Reply:
Hm, nice catch let me investigate!